### ISSUE TITLE
quotes: "avoidEscape" doesn't flag strings that escape quotes.

### ISSUE THREAD

**Issue by jpickwell** [eslint/eslint#10956]
**Tell us about your environment**

* **ESLint Version:** 5.6.1
* **Node Version:** 10.11.0
* **npm Version:** 6.4.1
* **yarn Version:** 1.10.1

**What parser (default, Babel-ESLint, etc.) are you using?** vue-eslint-parser

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
const standardConfig = require('eslint-config-standard/eslintrc.json')

const OFF = 'off'
const WARN = 'warn'
const ERROR = 'error'

module.exports = {
  extends: [
    'eslint:recommended',
    'plugin:flowtype/recommended',
    'plugin:lodash/recommended',
    'plugin:you-dont-need-lodash-underscore/compatible-warn',
    'plugin:vue/recommended',
    'standard'
  ],
  plugins: ['babel'],
  env: { commonjs: true },
  rules: {
    // disable native rules because of Babel plugin
    'new-cap': OFF,
    'no-unused-expressions': OFF,
    'object-curly-spacing': OFF,
    semi: OFF,
    'valid-typeof': OFF,

    'array-bracket-newline': ERROR,
    'array-bracket-spacing': ERROR,
    'array-callback-return': ERROR,
    'arrow-body-style': ERROR,
    'arrow-parens': ERROR,
    'babel/new-cap': standardConfig.rules['new-cap'],
    'babel/no-unused-expressions': standardConfig.rules['no-unused-expressions'],
    'babel/object-curly-spacing': standardConfig.rules['object-curly-spacing'],
    'babel/semi': standardConfig.rules.semi,
    'babel/valid-typeof': standardConfig.rules['valid-typeof'],
    'callback-return': ERROR,
    'computed-property-spacing': ERROR,
    'consistent-return': ERROR,
    'consistent-this': [ERROR, 'self'],
    'default-case': ERROR,
    'dot-notation': ERROR,
    'func-names': [ERROR, 'as-needed'],
    'global-require': ERROR,
    'guard-for-in': ERROR,
    'init-declarations': [ERROR, 'never'],
    'jsx-quotes': ERROR,
    'linebreak-style': ERROR,
    'lines-between-class-members': ERROR,
    'lodash/prefer-lodash-method': [
      ERROR,
      {
        ignoreMethods: ['isArray', 'replace', 'toLower', 'toUpper', 'trim']
      }
    ],
    'max-len': [ERROR, {
      code: 120,
      ignoreTrailingComments: true,
      ignoreUrls: true
    }],
    'max-statements-per-line': ERROR,
    'newline-per-chained-call': ERROR,
    'no-await-in-loop': ERROR,
    'no-confusing-arrow': [ERROR, { allowParens: true }],
    'no-div-regex': ERROR,
    'no-duplicate-imports': [ERROR, { includeExports: true }],
    'no-else-return': ERROR,
    'no-empty-function': ERROR,
    'no-extra-label': ERROR,
    'no-implicit-coercion': ERROR,
    'no-implicit-globals': ERROR,
    'no-lonely-if': ERROR,
    'no-loop-func': ERROR,
    'no-magic-numbers': [ERROR, { ignore: [-1, 0, 1] }],
    'no-mixed-requires': [ERROR, { grouping: true }],
    'no-negated-condition': ERROR,
    'no-nested-ternary': ERROR,
    'no-process-exit': ERROR,
    'no-prototype-builtins': ERROR,
    'no-script-url': ERROR,
    'no-shadow': ERROR,
    'no-useless-concat': ERROR,
    'no-var': ERROR,
    'no-warning-comments': [WARN, { terms: ['fixme', 'hack', 'todo', 'xxx'] }],
    'object-curly-newline': [ERROR, {
      ObjectExpression: {
        minProperties: 2,
        multiline: true
      },
      ObjectPattern: { multiline: true }
    }],
    'object-shorthand': ERROR,
    'operator-assignment': ERROR,
    'padding-line-between-statements': [
      ERROR,
      {
        blankLine: 'always',
        prev: ['const', 'let', 'var'],
        next: '*'
      },
      {
        blankLine: 'any',
        prev: ['const', 'let', 'var'],
        next: ['const', 'let', 'var']
      },
      {
        blankLine: 'always',
        prev: '*',
        next: 'return'
      }
    ],
    'prefer-arrow-callback': ERROR,
    'prefer-const': ERROR,
    'prefer-numeric-literals': ERROR,
    'prefer-rest-params': ERROR,
    'prefer-spread': ERROR,
    'prefer-template': ERROR,
    'quote-props': [ERROR, 'as-needed'],
    radix: ERROR,
    'require-await': ERROR,
    'semi-style': [ERROR, 'first'],
    'sort-imports': [ERROR, { ignoreCase: true }],
    'sort-vars': [ERROR, { ignoreCase: true }],
    'switch-colon-spacing': ERROR,
    'vars-on-top': ERROR,
    'vue/component-name-in-template-casing': [ERROR, 'kebab-case'],
    'vue/jsx-uses-vars': OFF,
    'vue/multiline-html-element-content-newline': ERROR,
    'vue/no-spaces-around-equal-signs-in-attribute': ERROR,
    'vue/no-v-html': OFF,
    // Disabled until https://github.com/vuejs/eslint-plugin-vue/issues/595 is fixed.
    'vue/require-prop-type-constructor': OFF,
    'wrap-regex': ERROR,
    'you-dont-need-lodash-underscore/assign': OFF
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
export default () => {
  const dq = "'\"'"
  const sq = '\''
  const tl = `'\`'`

  /* eslint-disable no-console */
  console.log(dq)
  console.log(sq)
  console.log(tl)
  /* eslint-enable no-console */
}
```

```bash
# eslint --cache --ext .js,.vue app/javascript
yarn lint
```

**What did you expect to happen?**

Expected the three string literals above to be flagged as errors.

The following shows how the strings could be written to avoid escaping.
```js
export default () => {
  const dq = `'"'`
  const sq = "'"
  const tl = "'`'"

  /* eslint-disable no-console */
  console.log(dq)
  console.log(sq)
  console.log(tl)
  /* eslint-enable no-console */
}
```

**What actually happened? Please include the actual, raw output from ESLint.**

No errors.

```
λ yarn lint
yarn run v1.10.1
$ yarn eslint app/javascript
$ eslint --cache --ext .js,.vue app/javascript
✨  Done in 1.58s.
```

**Comment by kaicataldo** [comment]
I believe this is working as intended. The `avoidEscape` option doesn't enforce avoiding escaping, it just allows for the quotes to not match the string option (`"single"`, `"double"`, etc.) in the case where it allows escaping to be avoided.

It sounds like you're looking for the [`no-useless-escape` rule](https://eslint.org/docs/rules/no-useless-escape).

**Comment by jpickwell** [comment]
`eslint-config-standard` has `no-useless-escape` set to `"error"`.

**Comment by kaicataldo** [comment]
@jpickwell I see. Having a little bit more time to look at things this morning, it looks to me like core doesn't have a way to enforce this. It falls in a bit of an awkward place between `no-useless-escape` and `quotes`, in that it is a legitimate escape (and so `no-useless-escape` doesn't warn on it), and that the `quotes` rule doesn't enforce using other types of quotes with the `avoidEscape` option - it just allows for them.

If we do update something here, I think that, responsibility-wise, it makes more sense for `no-useless-escape` to enforce this. That being said, `no-useless-escape` won't have knowledge about what types of quotes are allowed. 

I wonder if `avoidEscape` could be updated to have three options - something like `"never"`, `"allow"`, and `"enforce"`, or something. Thoughts?

**Comment by jpickwell** [comment]
The `no-useless-escape` rule's purpose seems very clear to me, and has nothing to do with this issue. That rule is meant to avoid escaping non-special characters.

As for your suggestion of `"never"`, `"allow"`, and `"enforce"`, would that map like this?
* `"never"`: `false`
* `"allow"`: `true`
* `"enforce"`: new feature

If that's the case, I'm good with that.

An obvious edge case that can't be handled by any rule would be a string containing all three types of quote characters. One type of quote character is going to need to be escaped.

Aside, I wish JavaScript had Heredoc strings. Especially Ruby's "squiggly" Heredoc which allows indentation.

**Comment by kaicataldo** [comment]
>As for your suggestion of `"never"`, `"allow"`, and `"enforce"`, would that map like this?
>* `"never"`: `false`
>* `"allow"`: `true`
>* `"enforce"`: new feature

That's what I was thinking - thanks for clarifying.
